### PR TITLE
shim-sev: handle SEV encryption

### PR DESCRIPTION
**Pull request by haraldh**

### CODE REVIEW

File: internal/shim-sev/src/start.S [hunk 1]
```diff
@@ -25,6 +25,52 @@
...
 _start:
+.intel_syntax noprefix
+CheckSevFeature:
+    // Check if we have a valid (0x8000_001F) CPUID leaf
+    mov       eax, 0x80000000
+    cpuid
+
+    // This check should fail on Intel or Non SEV AMD CPUs. In future if
+    // Intel CPUs supports this CPUID leaf then we are guranteed to have exact
+    // same bit definition.
```
Comment (npmccallum) [inline internal/shim-sev/src/start.S:36]: I'm skeptical about this claim.

**Review by npmccallum — CHANGES_REQUESTED** [review]
Two larger concerns:

1. The amount of asm has grown significantly. We now have a megafunction written in assembly that is very hard to follow. Can this be broken up?

2. I thought we agreed on making the C-bit a constant with an assert if it changes.

File: internal/shim-sev/src/start.S (hunk 1 above)
Comment (haraldh) [reply internal/shim-sev/src/start.S:36]: Ooops, shamelessly copied.

Author: Brijesh Singh <brijesh.singh@amd.com>

https://github.com/AMDESE/ovmf/commit/e60af8a1ebb15bfcbf2ecc4afb6cf35084c847aa

**Comment by haraldh** [comment]
> Two larger concerns:
> 
>     1. The amount of asm has grown significantly. We now have a megafunction written in assembly that is very hard to follow. Can this be broken up?
> 
>     2. I thought we agreed on making the C-bit a constant with an assert if it changes.

1. IMHO, nothing before  `call    _dyn_reloc` can be made into rust safely
2. wasn't that big of a deal and a hardcoded version wouldn't change anything

**Comment by haraldh** [comment]
>     1. The amount of asm has grown significantly. We now have a megafunction written in assembly that is very hard to follow. Can this be broken up?

the sgx `start.S` has an equal amount of lines

File: internal/shim-sev/src/start.S (hunk 1 above)
Comment (haraldh) [reply internal/shim-sev/src/start.S:36]: want me to remove the comments?

**Review by npmccallum — APPROVED** [review]

**Review by connorkuehl — APPROVED** [review]
